src/firmware/shared/math/matrix.c

```diff
@@ -1,103 +1,130 @@
+    Matrix_t* return_matrix = shared_matrix_create(A->n_rows, B->n_cols);
+    for (unsigned int i = 0; i < B->n_cols; i++)
     {
-        for (j = 0; j < A.n_rows; j++)
+        for (unsigned int j = 0; j < A->n_rows; j++)
         {
-            C.rows[j][i] = 0.0f;
-            for (k = 0; k < A.n_cols; k++)
+            return_matrix->rows[j][i] = 0.0f;
+            for (unsigned int k = 0; k < A->n_cols; k++)
             {
-                C.rows[j][i] += A.rows[j][k] * B.rows[k][i];
+                return_matrix->rows[j][i] += A->rows[j][k] * B->rows[k][i];
             }
         }
     }
-    return C;
+    return return_matrix;
 }
 
-/**
- * Rotates the given vector by using the given rotation matrix.
- *
- * @param vector a 2D vector to rotate
- * @param rotation_matrix the rotation operator
- */
-void do_rotation(float vector_array[2], Matrix rotation_matrix)
+Matrix_t* shared_matrix_transpose(Matrix_t* in_matrix)
 {
-    Matrix vector     = create_matrix(2, 1);
-    vector.rows[0][0] = vector_array[0];
-    vector.rows[1][0] = vector_array[1];
-    Matrix rotated    = matmul(rotation_matrix, vector);
-    vector_array[0]   = rotated.rows[0][0];
-    vector_array[1]   = rotated.rows[1][0];
-    free_matrix(vector);
-    free_matrix(rotated);
-    free_matrix(rotation_matrix);
+    Matrix_t* out_matrix = shared_matrix_create(in_matrix->n_cols, in_matrix->n_rows);
+    for (unsigned int i = 0; i < in_matrix->n_rows; i++)
+    {
+        for (unsigned int j = 0; j < in_matrix->n_cols; j++)
+        {
+            out_matrix->rows[j][i] = in_matrix->rows[i][j];
+        }
+    }
+    return out_matrix;
 }
 
-void rotate_axis_2D(float vector[2], const float unit_vector[2])
+void shared_matrix_setValueAtIndex(unsigned int row, unsigned int column, float value,
+                                   Matrix_t* matrix)
 {
-    Matrix rotation_matrix     = create_matrix(2, 2);
-    rotation_matrix.rows[0][0] = unit_vector[0];
-    rotation_matrix.rows[0][1] = unit_vector[1];
-    rotation_matrix.rows[1][0] = -unit_vector[1];
-    rotation_matrix.rows[1][1] = unit_vector[0];
-    do_rotation(vector, rotation_matrix);
+    assert(row <= matrix->n_rows && row >= 1);
+    assert(column <= matrix->n_cols && column >= 1);
+
+    // If the row or column are out of the range of the array then return.
+    // It is likely safer to do nothing than segfault
+    if (row > matrix->n_rows || column > matrix->n_cols)
+    {
+        return;
+    }
+    matrix->rows[row - 1][column - 1] = value;
+}
+
+float shared_matrix_getValueAtIndex(unsigned int row, unsigned int column,
+                                    Matrix_t* matrix)
+{
+    assert(row <= matrix->n_rows && row >= 1);
+    assert(column <= matrix->n_cols && column >= 1);
+
+    // If the row or column are out of the range of the array then return.
+    // It is likely safer to do nothing than segfault
+    if (row > matrix->n_rows || column > matrix->n_cols || row < 1 || column < 1)
```

Comment:
row and column are already asserted to be >= 1 so the last two checks are redundant